Mathlib/MeasureTheory/Measure/Tight.lean

```diff
@@ -0,0 +1,483 @@
+
+## Main definitions
+* `MeasureTheory.Measure.IsSeparablySupported`: A measure `μ` is separably supported if there is a
+  separable set `S` such that `μ Sᶜ = 0`.
+* `MeasureTheory.Measure.IsPretight`: A measure `μ` is pre-tight if for all `0 < ε`, there exists
+  `K` totally bounded such that `μ Kᶜ ≤ ε`.
+* `MeasureTheory.Measure.IsTight`: A measure `μ` is tight if for all `0 < ε`, there exists `K`
+  compact such that `μ Kᶜ ≤ ε`. This is defined in terms of filters.
+  `iff_exists_isCompact_measure_compl_le` establishes equivalence with the usual definition.
+
+## Main statements
+
+* `of_isSeparableSpace_completeSpace`: Ulam's tightness theorem: a finite measure on a complete
+  separable metric space is tight.
+* `of_isSeparablySupported_completeSpace`: A strengthening of Ulam's tightness theorem: a finite,
+  separably supported measure on a complete metric space is tight.
+* `of_innerRegular`: Every finite, inner-regular measure on a T2 space is tight.
+
+## Notation
+
+
+## Implementation details
+
+
+-/
+open Topology Filter Uniformity Uniform MeasureTheory Set
+open scoped ENNReal
+
+namespace MeasureTheory.Measure
+
+variable {α ι : Type*} [MeasurableSpace α] {μ : Measure α}
+
+/-- For a finite measure `μ`, we can extract from a countable cover that has full measure, a finite
+cover of accumulated sets that has `ε`-almost full measure. -/
+lemma exists_sub_le_measure_accumulate [IsFiniteMeasure μ] (K : ℕ → Set α)
+    (h : μ (⋃ n, K n) = μ Set.univ) {ε : ENNReal} (hε : ε > 0) :
+    ∃ n, μ Set.univ - ε ≤ μ (Set.Accumulate K n) := by
+  have : Filter.Tendsto (μ ∘ Set.Accumulate K) Filter.atTop (nhds (μ (⋃ n, Set.Accumulate K n))) :=
+    MeasureTheory.tendsto_measure_iUnion Set.monotone_accumulate
+  rw [ENNReal.tendsto_atTop (measure_ne_top μ (⋃ n, Accumulate K n)), Set.iUnion_accumulate] at this
+  obtain ⟨N, hN⟩ := this ε hε
+  use N
+  simp_all only [Function.comp_apply, mem_Icc, le_refl]
+
+/-- For a finite measure `μ`, we can extract from a countable cover that has full measure, a finite
+cover of accumulated sets for which the complement has measure below `ε`. -/
+lemma exists_sub_le_measure_accumulate_compl [IsFiniteMeasure μ] (K : ℕ → Set α)
+    (hKmeas : ∀ n, MeasurableSet (K n)) (h : μ (⋃ n, K n) = μ Set.univ) {ε : ENNReal} (hε : ε > 0) :
+    ∃ n, μ ((Set.Accumulate K n)ᶜ) ≤ ε := by
+  obtain ⟨n, hn⟩ := exists_sub_le_measure_accumulate K h hε
+  use n
+  have hK2 : MeasurableSet (Set.Accumulate K n) := by
+    rw [Set.Accumulate]
+    measurability
+  rw [measure_compl hK2 (measure_ne_top μ _)]
+  exact tsub_le_iff_tsub_le.mp hn
+
+/-- For a finite measure `μ`, we can construct a dense sequence such that for any radius, we can
+cover all but a set of measure below `ε`. -/
+lemma exists_denseRange_sub_le_measure_univ_of_separableSpace [PseudoMetricSpace α]
+    [TopologicalSpace.SeparableSpace α] [Nonempty α] (μ : Measure α) [IsFiniteMeasure μ] :
```

Comment:
This formulation is really strange: what the lemma seems to be saying is: I can construct a clever sequence `u` with a dense range such that for all positive epsilon, ...
But what is true is rather: whatever sequence u with a dense range, then ...
So you should rather start from a sequence with a dense range, and prove your conclusion for it. In the application of the lemma, pick any sequence using separability and apply the lemma to this sequence. In this way, it is clearer that separability and nonemptyness of alpha are not really relevant here. 

Even better, it points you to a better lemma: if `u` is a sequence such that `μ ((closure (range u))ᶜ) = 0`, then the conclusion of your lemma holds, with the same proof. The advantage of this new lemma (that should probably replace the current `exists_denseRange_sub_le_measure_univ_of_separableSpace`) is that you can avoid all discussion of subtypes in the tedious proof of `of_isSeparablySupported_on_countablyGenerated_uniformSpace` below (i.e., don't prove it first assuming the space is separable and then restricting to a subtype, but instead give a direct proof).